### ISSUE TITLE
[Question] Who (which component) is responsible for unloading JIT?

### ISSUE THREAD

**Issue by parjong** [dotnet/coreclr#12031]
I discovered that ``jitShutdown`` (in ``ee_il_dll.cpp``) is never called on Linux while trying to profile JIT using ``COMPlus_JitTimeLogFile``.

It seems that ``DllMain`` in ``ee_il_dll.cpp`` should invoke ``jitShutdown`` when ``DLL_PROCESS_DETACH`` is notified, but ``DLL_PROCESS_DETACH`` is never notified.

So, the question is which component should be fixed to support graceful JIT unloading?

**Comment by janvorli** [comment]
@parjong, can you please describe a scenario when you would need the jitShutdown to be called? The DLL_PROCESS_DETACH is notified on Unix only when a shared library is released using the FreeLibrary API.
When the CoreCLR app exits, aborts or crashes on Unix, we don't try to deliver these notifications.

**Comment by parjong** [comment]
``Compiler::compShutdown`` seems to reports some performance statistics. I would like to see this statistics. 

I guess that I need this information when app normally exits or aborted by some exception, but does not need for other cases (aborted by assertion or crashes)

It would be good to have this even for those cases if it is not so difficult.

**Comment by janvorli** [comment]
@parjong maybe we could add UnloadJIT function to the EEJitManager class, let it call FreeLibrary on the m_JITCompiler and then call this function e.g. from CorHost2::Stop. That should result in the DllMain being called with DLL_PROCESS_DETACH. That would not work if the process called "exit".
The other option would be to add new module cleanup function to module.cpp and call it from PALCommonCleanup. This function would loop over all loaded modules and free them or at least call their DllMain with DLL_PROCESS_DETACH. I am not sure though whether there are any possible bad consequences or not.

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib In case you have a better way to solve @parjong's problem.

**Comment by jkotas** [comment]
> maybe we could add UnloadJIT function to the EEJitManager class

This method exist already. It is called `ProcessShutdownWork`: https://github.com/dotnet/coreclr/blob/b2164d274f13023e5546fab849f6f15381ba4475/src/inc/corjit.h#L427 . The VM calls it during graceful process shutdown to give JIT opportunity to do its shutdown work.

**Comment by parjong** [comment]
@janvorli I tried to invoke ``LOADCallDllMain`` in ``PALCommonCleanup`` (I guess that this is the 2nd approach that you suggested), but the code itself has the following assert:
```
    case DLL_PROCESS_DETACH:
        ASSERT("got called with DLL_PROCESS_DETACH parameter! Why?\n");
        InLoadOrder = FALSE;
        break;
```

Is that prohibited to call this method? I guess that the loop at the end of ``LOADCallDllMain`` loops over all loaded modules (that is currently needed).

**Comment by parjong** [comment]
@jkotas Thank you for comment. I'll take a look.

**Comment by janvorli** [comment]
@parjong yes, that's what I meant, I have forgotten we have the LOADCallDllMain function. The assert for DLL_PROCESS_DETACH is there because it was never used and it was not expected to happen. 
However, I would prefer the other way (but instead of calling the new UnloadJIT from the CorHost2::Stop, we would call it from the place that @jkotas  has suggested.
@jkotas the issue is that the ProcessShutdownWork method is on ICorJitCompiler while the handle of the JIT library that we need to pass it is stored in the EEJitManager and the JIT is loaded using `EEJitManager::LoadJIT()`. That's why I hav suggested to add the UnloadJIT method to the EEJitManager so that the JIT loading and unloading is encapsulated in the same object.
If we did it the way I've suggested, we would call the UnloadJIT from the `CEEInfo::JitProcessShutdownWork` after the calls to the ProcessShutdownWork.
